DOCSP-48915: rocket crud web app tutorial (docs+) (#191)
mongodb/docs-rust: source/fundamentals/crud/web-app-tutorial.txt

Quick Start guide. Save your connection string in a safe location to use
later in the tutorial.

Steps
-----

.. procedure::
   :style: connected

   .. step:: Create a {+language+} project.

      Run the following commands to create and enter the
      ``rust-crud-app`` project directory:

      .. code-block:: bash

         cargo new rust-crud-app
         cd rust-crud-app

   .. step:: Install the {+driver-short+}.

      Run the following command to add the {+driver-short+}:

      .. code-block:: shell

         cargo add mongodb

      Verify that your ``Cargo.toml`` file contains the following entry
      for the {+driver-short+}:

      .. code-block:: toml

         [dependencies]
         mongodb = "{+version+}"

   .. step:: Select the asynchronous runtime.

      When using the {+driver-short+}, you must select either the
      synchronous or asynchronous runtime. This tutorial uses the
      asynchronous runtime, which is more suitable for building APIs.

      The driver runs by default with the async ``tokio`` runtime.

      To learn more about the available runtimes, see the
      :ref:`rust-runtimes` guide.

   .. step:: Insert sample data.

>       In the Atlas UI, create a ``bread`` database, and

Author: 1. Will separate these into two steps

      flag for ``mongodb``.

   .. step:: Configure Rocket.

      To configure Rocket to use the ``bread`` database, create a
      file called ``Rocket.toml`` in your project root. Rocket looks for
      this file to read configuration settings. You can also store your
      MongoDB connection string in this file.

      Paste the following configuration into ``Rocket.toml``:

      .. code-block:: toml

         [default.databases.db]
         url = "<connection string>"

      To learn more about configuring Rocket, see `Configuration
      <https://rocket.rs/guide/v0.5/configuration/#configuration>`__ in
      the Rocket documentation.

   .. step:: Learn about the app structure.

      Before you begin writing the API, learn about the structure of a
      simple Rocket app.

      The following diagram demonstrates the file structure of a Rocket
      app that explains where different components are stored:

      .. code-block:: text

         .
         ├── Cargo.lock      # Dependency info
         ├── Cargo.toml      # Project and dependency info
         ├── Rocket.toml     # Rocket configuration
         └── src             # Directory for all app code

>              ├── db.rs       # Establish database connection
>              ├── main.rs     # Start the web app

Reviewer: [s] Consistency with the other labels:
```suggestion
             ├── db.rs       # Establishes database connection
             ├── main.rs     # Starts the web app
```

             ├── models.rs   # Organizes data
             └── routes.rs   # Stores API routes

      Create the ``src`` directory and its files, which can be blank at

>       this step.

Reviewer: Mixing the reference content with an instructional step is a little confusing here. Maybe this line can specify that "its files" means the same files outlined in the above diagram? Or can update the intro sentence for the diagram to explain that this is the file structure the tutorial's app uses. The current wording sounds like it's just for reference/explanation.

   .. step:: Set up the database connection.

      Paste the following code into the ``db.rs`` file.

      .. code-block:: rust

         use rocket_db_pools::{mongodb::Client, Database};

         #[derive(Database)]
         #[database("db")]
         pub struct MainDatabase(Client);

      You must also attach the database struct to your Rocket
      instance. In ``main.rs``, initialize the database and
      attach it, as shown in the following code:

>       .. code-block:: rust

Reviewer: [s] For future maintenance I wonder if moving the code blocks to includes would be helpful as future code testing is implemented?

Author: The reason I didnt use includes is because the tutorial creates and replaces content, as well as the fact that I thought creating all the app files seemed superfluous for just a few includes. I can create just the routes.rs file though?

      Before you write the remaining routes, add the routes to the main
      launch function for Rocket.

      In ``main.rs``, replace the arguments to ``mount()`` so that the
      file resembles the following code:

      .. code-block:: rust

         mod db;
         mod models;
         mod routes;

         use rocket::{launch, routes};
         use rocket_db_pools::Database;

         #[launch]
         fn rocket() -> _ {
             rocket::build().attach(db::MainDatabase::init()).mount(
                 "/",
                 routes![
                     routes::index,
                     routes::get_recipes,
                     routes::create_recipe,
                     routes::update_recipe,
                     routes::delete_recipe,
                     routes::get_recipe
                 ],
             )
         }

   .. step:: Implement error handling and responses.

      In your app, you must implement error handling and custom
      responses to deal with unexpected outcomes of your CRUD
      operations.

      First, install the ``serde_json`` crate by running the following
      command:

      .. code-block:: bash

         cargo add serde_json

>       Then, modify the ``get_recipes()`` route shown to avoid panics and

Reviewer: ```suggestion
      Then, modify the ``get_recipes()`` route from the previous step to avoid panics and
```

      return custom HTTP status codes. You can leverage Rocket’s
      ``status::Custom`` structs, which allow you to specify the HTTP
      status code and any custom data to return.

      Modify the return type of ``get_recipes()`` to be

>       ``status::Custom<Json<Value>>``. This lets the {+language+} compiler know

Reviewer: [s] May be helpful to include a snippet here of what get_recipes() should look like with this status added, maybe with the line highlighted?

         ) -> status::Custom<Json<Value>> {
             if let Ok(res) = db
                 .database("bread")
                 .collection::<Recipe>("recipes")
                 .insert_one(data.into_inner(), None)
                 .await
             {
                 if let Some(id) = res.inserted_id.as_object_id() {
                     return status::Custom(
                         Status::Created,
                         Json(json!({"status": "success", "message": format!("Recipe ({}) created successfully", id.to_string())})),
                     );
                 }
             }
      
             status::Custom(
                 Status::BadRequest,
                 Json(json!({"status": "error", "message":"Recipe could not be created"})),
             )
         }

      Read
      ~~~~

      When you attempt to read data from MongoDB, there are two possible
      outcomes:

      - Return the vector of matching documents.
      - Return an empty vector, because there are no matching documents
        or because an error occurred.

      Because of these expected outcomes, replace your ``get_recipes()``
      route with the following code:
      

>       .. code-block:: rust

Reviewer: [q] So does this mean the ``status`` described above isn't needed? I don't see it in this snippet.

Author: Yes, its kind of weird that the code makes you redo the step you just completed. I can just remove the preceding step and reword it.

      
         #[get("/recipes", format = "json")]
         pub async fn get_recipes(db: Connection<MainDatabase>) -> Json<Vec<Recipe>> {
             let recipes = db
                 .database("bread")
                 .collection("recipes")
                 .find(None, None)
                 .await;
      
             if let Ok(r) = recipes {
                 if let Ok(collected) = r.try_collect::<Vec<Recipe>>().await {
                     return Json(collected);
                 }
             }
      
             return Json(vec![]);
         }

      Other Operations
      ~~~~~~~~~~~~~~~~

      You can copy the ``get_recipe()``, ``update_recipe()``, and
      ``delete_recipe()`` routes from the :github:`routes.rs
      <latonis/mongodb-api-rs/blob/main/src/routes.rs>` file in the
      sample app repository.

   .. step:: Test routes to perform CRUD operations.

>       First, start your application by running the following command in

Reviewer: Since this section starts off with "First," it seems like the next steps should use some kind of sequential phrasing as well (e.g. Next, Then etc).

      your terminal:

      .. code-block:: bash
      
         cargo run
      
      In another terminal window, run the following command to test the
      ``create_recipe()`` route:
      
      .. code-block:: bash
      
         curl -v --header "Content-Type: application/json" --request POST --data '{"title":"simple bread recipe","ingredients":["water, flour"], "temperature": 250, "bake_time": 120}' http://127.0.0.1:8000/recipes

>       Run the following command to test the ``get_recipes()`` route:

Reviewer: [s] Could this section also include an example of what gets returned?

      
      .. code-block:: bash
      
         curl -v --header "Content-Type: application/json" --header "Accept: application/json" http://127.0.0.1:8000/recipes/

      Run the following command to test the ``delete_recipe()`` route.
      Replace the ``<id>`` placeholder with a known ``_id`` value from
      your collection, which might resemble ``68484d020f561e78c03c7800``:
      
      .. code-block:: bash
      
         curl -v --header "Content-Type: application/json" --header "Accept: application/json" --request DELETE http://127.0.0.1:8000/recipes/<id>

> Conclusion

Reviewer: [s] Can potentially add a Resources section here with a complete list of links to the relevant documentation/concepts.

      synchronous or asynchronous runtime. This tutorial uses the
      asynchronous runtime, which is more suitable for building APIs.

      The driver runs with the async ``tokio`` runtime by default.

      To learn more about the available runtimes, see the
      :ref:`rust-runtimes` guide.

   .. step:: Create a database and collection in the Atlas UI.

      Access the the Atlas UI, then select the :guilabel:`Collections` tab

>       in your cluster settings. Select the :guilabel:`+ Create Database`
>       button, then create a database called ``bread``, and within it, a
>       collection called ``recipes``.

Reviewer: [s] Similar to the above: 
```suggestion
      in your cluster settings. Select the :guilabel:`+ Create Database`
      button; then create a database called ``bread``, and within it, a
      collection called ``recipes``.
```
